src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,188 @@
 package core.basesyntax;
 
+import java.util.Collection;
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T item, Node<T> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        final Node<T> l = tail;
+        final Node<T> newNode = new Node<>(l, value, null);
+        tail = newNode;
+
+        if (l == null) {
+            head = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddingIndex(index);
+
+        if (index == size) {
+            add(value);
+        } else {
+            linkBefore(value, getNode(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        addAll(size, list);
+    }
+
+    private void addAll(int index, Collection<? extends T> collection) {
+        checkAddingIndex(index);
+        T[] objects = (T[]) collection.toArray();
+
+        if (index == size) {
+            for (T elem : objects) {
+                add(elem);
+            }
+        } else {
+            for (T element : objects) {
+                linkBefore(element, getNode(index));
+                index++;
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNode(index).item;
     }
```

Comment:
This is code duplication. check Index inside getNode(index)
![image](https://user-images.githubusercontent.com/20326858/215869753-0a376352-d17c-453e-a1c2-f5f1e9cc0df3.png)

Reply:
Now, I completely understand that

> This is code duplication. check Index inside getNode(index)
> ![image](https://user-images.githubusercontent.com/20326858/215869753-0a376352-d17c-453e-a1c2-f5f1e9cc0df3.png)